### PR TITLE
CEM planner for MBRL

### PR DESCRIPTION
**Pull request by Haichao-Zhang**

### CODE REVIEW

File: alf/algorithms/planning_algorithm.py [hunk 1]
```diff
@@ -269,3 +275,107 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
 
     def after_update(self, training_info):
         pass
+
+
+@gin.configurable
+class CEMPlanAlgorithm(RandomShootingAlgorithm):
+    """CEM-based planning method.
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py:241]: Add a reference and/or briefly describe the algorithm.

File: alf/optimizers/traj_optimizers.py [hunk 2]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
         # solutions [B, pop_size, sol_dim] -> [B, sol_dim]
         solution = solutions[(torch.arange(batch_size), min_ind)]
         return solution
+
+
+class CEMOptimizer(TrajOptimizer):
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py:83]: Describe what this optimizer do.

File: alf/optimizers/traj_optimizers.py [hunk 3]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+
+        i = 0
+        pop_mean = init_mean
+        pop_var = init_var
+
+        # [B, population, solution_dim]
+        trunc_normal_samples = torch.zeros(batch_size, self._population_size,
+                                           self._solution_dim)
+
+        while i < self._max_iter_num and pop_var.view(
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: `pop_var.view(-1).max()` => `pop_var.max()`

File: alf/algorithms/planning_algorithm.py (hunk 1 above)
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py:241]: need to note that this is not thread-safe due to the use of self._prev_solution

File: alf/algorithms/planning_algorithm.py [hunk 4]
```diff
@@ -269,3 +275,107 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
...
+                time_step.step_type == StepType.FIRST)
+
+        self._plan_optimizer.set_cost(self._calc_cost_for_action_sequence)
+
+        init_mean = self._prev_solution.unsqueeze(1)
+        opt_action = self._plan_optimizer.obtain_solution(
+            time_step,
+            state,
+            init_mean=init_mean,
+            init_var=torch.ones_like(init_mean) * 1)
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py]: This seems to assume that the actions are in certain range? Should make this proportional to the action bound.

File: alf/optimizers/traj_optimizers.py [hunk 5]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+        batch_size = init_obs.shape[0]
+
+        if init_mean is None:
+            # [B, 1, solution_dim]
+            init_mean = torch.zeros(batch_size, 1, self._solution_dim)
+        else:
+            assert init_mean.shape == (batch_size, 1, self._solution_dim)
+
+        if init_var is None:
+            init_var = torch.ones(batch_size, 1, self._solution_dim)
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: Instead of being fixed, it should be derived from `upper_bound` and `lower_bound`

File: alf/optimizers/traj_optimizers.py [hunk 6]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+                trunc_normal_samples)
+            samples = trunc_normal_samples * torch.sqrt(
+                constrained_var) + pop_mean
+
+            costs = self.cost_function(time_step, state, samples.clone())
+
+            # select elite set from the population
+            ind = torch.topk(-costs, self._elite_size)[1]
+            # samples: [batch, population, horizon]
+            elites = samples.index_select(1, ind.view(-1))
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: This is not correct when batch_size is not 1.
```
elites = samples[torch.arange(batch_size).unsqueeze(-1), ind]
```

File: alf/optimizers/traj_optimizers.py [hunk 7]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+                 min_var=1e-5):
+        """Creates a CEM Optimizer
+
+        Args:
+            solution_dim (int): the dimensionality of the problem space
+            population_size (int): the number of candidate solutions to be
+                sampled at every iteration
+            upper_bound (float|Tensor): upper bounds for elements in solution
+            lower_bound (float|Tensor): lower bounds for elements in solution
+            elite_size (int): the number of elites selected in each round
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: Need to explain how elites are used.

File: alf/algorithms/planning_algorithm.py [hunk 8]
```diff
@@ -269,3 +275,107 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
...
+
+        # [B, pop_size=1, horizon * action_dim] -> [B, horizon, action_dim]
+        opt_action = torch.reshape(
+            opt_action,
+            [opt_action.shape[0], self._planning_horizon, self._num_actions])
+
+        # [B, horizon, action_dim]
+        self._prev_solution = torch.cat(
+            (opt_action[:, 1:], torch.zeros_like(opt_action[:, 0:1])), 1)
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py]: 0 might be out of action bound. Perhaps use mean? `opt_action.mean(dim=(0, 1), keepdim=True).expand(batch_size, -1, -1)`

File: alf/optimizers/traj_optimizers.py [hunk 9]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+            trunc_normal_samples = alf.initializers.trunc_normal_(
+                trunc_normal_samples)
+            samples = trunc_normal_samples * torch.sqrt(
+                constrained_var) + pop_mean
+
+            costs = self.cost_function(time_step, state, samples.clone())
+
+            # select elite set from the population
+            ind = torch.topk(-costs, self._elite_size)[1]
+            # samples: [batch, population, horizon]
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: [batch, population, solution_dim]

File: alf/optimizers/traj_optimizers.py [hunk 10]
```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+            distance_to_ub = self._upper_bound - pop_mean
+
+            # compute the constrained var based on the computed distance
+            constrained_var = torch.min(
+                torch.min((distance_to_lb / 2.0)**2, (distance_to_ub / 2.0)
+                          **2), pop_var)
+            constrained_var = constrained_var.clamp(min=self._min_var)
+
+            trunc_normal_samples = alf.initializers.trunc_normal_(
+                trunc_normal_samples)
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: It's better to modify `trunc_normal_()` to handle vector form of mean, std, a, b.

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/algorithms/planning_algorithm.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py:241]: Added comments and reference

File: alf/optimizers/traj_optimizers.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py:83]: Added description and reference.

File: alf/optimizers/traj_optimizers.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: changed

File: alf/algorithms/planning_algorithm.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Changed to adaptive to action bound.

File: alf/optimizers/traj_optimizers.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: Changed to be adaptive to action bound.

File: alf/optimizers/traj_optimizers.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: Corrected.

File: alf/optimizers/traj_optimizers.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: Added explanation.

File: alf/algorithms/planning_algorithm.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Switched to use mean.

File: alf/optimizers/traj_optimizers.py (hunk 9 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: Updated.

File: alf/optimizers/traj_optimizers.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: Leave ```trunc_normal_``` unchanged for now as it is a pytorch function (copied from a higher version). Will think about proper ways to deal with this issue.

File: alf/optimizers/traj_optimizers.py [hunk 11]
```diff
@@ -89,3 +91,133 @@ def obtain_solution(self, time_step: TimeStep, state):
...
+
+        while i < self._max_iter_num and pop_var.max() > self._epsilon:
+            # compute the distance to lower and upper bound
+            distance_to_lb = pop_mean - self._lower_bound
+            distance_to_ub = self._upper_bound - pop_mean
+
+            # compute the constrained var based on the computed distance
+            constrained_var = torch.min(
+                torch.min((distance_to_lb / 2.0)**2, (distance_to_ub / 2.0)
+                          **2), pop_var)
```
Comment (emailweixu) [inline alf/optimizers/traj_optimizers.py]: This will make the bound at least 2 standard deviation away from the mean. It can make the convergence slow if the desired solution is close the bound.

File: alf/algorithms/planning_algorithm.py [hunk 12]
```diff
@@ -269,3 +282,126 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
...
+    trajectory by minimizing a given cost function. The optimized action
+    trajectory is termed as a 'plan' which can be used by other components
+    such as a MPC-based controller. This has been used by some MBRL works
+    such as `Deep Reinforcement Learning in a Handful of Trials using
+    Probabilistic Dynamics Models <https://arxiv.org/abs/1805.12114>`_
+
+    To speedup, when possible, we have used the plan obtained at the previous
+    time step to initialize the the mean of the plan distribution at the current
+    time step, after proper shifting and padding. This, however, makes the
+    current implementation no longer thread-safe.
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py]: Perhaps we can use state to handle this to make it thread-safe

Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Good point. Will try to switch to state.

File: alf/optimizers/traj_optimizers.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/optimizers/traj_optimizers.py]: Changed to using Normal as generation distribution. Performs similar to TruncNormal but computation speed is much faster.

File: alf/algorithms/planning_algorithm.py (hunk 12 above)
Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Now switched to use state for passing the previous plan.

File: alf/algorithms/planning_algorithm.py [hunk 13]
```diff
@@ -269,3 +283,116 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
...
+                                               "before planning")
+
+        assert self._dynamics_func is not None, ("specify dynamics function "
+                                                 "before planning")
+
+        batch_size = time_step.observation.shape[0]
+
+        self._plan_optimizer.set_cost(self._calc_cost_for_action_sequence)
+
+        prev_plan = state.planner.prev_plan
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py]: one problem of this is that the first prev_plan is all zero. Will get a bad init_mean from it.

Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Now add back the self-defined solution resetting which is more adaptive to solution bounds.

File: alf/algorithms/planning_algorithm.py [hunk 14]
```diff
@@ -269,3 +283,130 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
...
+        # [B, horizon, action_dim] -> [B, horizon*action_dim]
+        prev_solution = prev_plan.reshape(prev_plan.shape[0], -1)
+
+        if self._default_solution is None:
+            self._default_solution = torch.ones(batch_size, self._solution_size) \
+                    * (self._upper_bound + self._lower_bound) / 2.
+
+        # reset to default solution at the beginning of an episode
+        prev_solution = common.reset_state_if_necessary(
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py]: The following is thread-safe:
```python
prev_solution = prev_solution.clone()
prev_solution[time_step.step_type==StepType.FIRST] = torch.full(self._solution_size, (self._upper_bound + self._lower_bound) / 2.)
```

File: alf/algorithms/mbrl_algorithm.py [hunk 15]
```diff
@@ -174,8 +174,8 @@ def _calc_step_reward(self, obs, action, state: MbrlState):
     def _predict_with_planning(self, time_step: TimeStep, state,
                                epsilon_greedy):
         # full state in
-        action = self._planner_module.generate_plan(time_step, state,
-                                                    epsilon_greedy)
+        action, state = self._planner_module.generate_plan(
+            time_step, state, epsilon_greedy)
```
Comment (emailweixu) [inline alf/algorithms/mbrl_algorithm.py]: pass in  `state.planner` instead as planner_module should not know about the structure of `state` 
and generate_plan => predict_plan

**Comment by Haichao-Zhang** [comment]
> Should pass `state.planner` to `generate_plan()` instead of `state`.
> Also feel it is better to change generate_plan to predict_step to be consistent with other places.

In the current design, ```generate_plan``` takes full ```MbrlState``` as input state. The reason is that
the cost function to be optimized over in the planner requires state information beyond `state.planner` in order to be calculated. For example, the  `state.dynamics` is needed in order to compute the next predicted observation.

Changed the name of function `generate_plan` to `predict_plan `.

File: alf/algorithms/mbrl_algorithm.py (hunk 15 above)
Comment (Haichao-Zhang) [reply alf/algorithms/mbrl_algorithm.py]: I agree that planner_module should not know about the structure of state. Currently there are still some flaws in the interface design.

Comment (Haichao-Zhang) [reply alf/algorithms/mbrl_algorithm.py]: Now changed name to predict_plan

File: alf/algorithms/planning_algorithm.py (hunk 14 above)
Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Corrected

File: alf/algorithms/mbrl_algorithm.py [hunk 16]
```diff
@@ -174,8 +174,8 @@ def _calc_step_reward(self, obs, action, state: MbrlState):
     def _predict_with_planning(self, time_step: TimeStep, state,
                                epsilon_greedy):
         # full state in
-        action = self._planner_module.generate_plan(time_step, state,
-                                                    epsilon_greedy)
+        action, state = self._planner_module.predict_plan(
+            time_step, state, epsilon_greedy)
```
Comment (emailweixu) [inline alf/algorithms/mbrl_algorithm.py]: For the planner module, `state.planner` is different from `state.dynamics` or `state.reward`. `state.dynamics` and `state.reward` are more like an input. While `state.planner` is indeed a state.

Comment (Haichao-Zhang) [reply alf/algorithms/mbrl_algorithm.py]: I have refactored the APIs for resolving the issue.

File: alf/algorithms/planning_algorithm.py [hunk 17]
```diff
@@ -107,24 +108,18 @@ def train_step(self, time_step: TimeStep, state: PlannerState):
...
-        """Set the dynamics function for planning
-        Args:
-            dynamics_func (Callable): reward function to be used for planning.
-            dynamics_func takes (time_step, state) as input and returns
-            next_time_step (TimeStep) and the next_state
-        """
-        self._dynamics_func = dynamics_func
-
-    def generate_plan(self, time_step: TimeStep, state, epsilon_greedy):
+    def predict_plan(self, time_step: TimeStep, state, epsilon_greedy):
```
Comment (emailweixu) [inline alf/algorithms/planning_algorithm.py]: comment for `state` should be updated.

Comment (Haichao-Zhang) [reply alf/algorithms/planning_algorithm.py]: Comments updated.

**Review by emailweixu — APPROVED** [review]
